compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala

```diff
@@ -920,17 +943,90 @@ abstract class GenJSCode extends plugins.PluginComponent
 
     // Static initializers -----------------------------------------------------
 
-    private def genStaticInitializerLoadingModule(sym: Symbol)(
+    private def genStaticInitializerWithStats(stats: js.Tree)(
         implicit pos: Position): js.MethodDef = {
       js.MethodDef(
           static = true,
           js.Ident(ir.Definitions.StaticInitializerName),
           Nil,
           jstpe.NoType,
-          Some(genLoadModule(sym)))(
+          Some(stats))(
           OptimizerHints.empty, None)
     }
 
+    private def genRegisterReflectiveInstantiation(sym: Symbol)(
+        implicit pos: Position): List[js.Tree] = {
+      if (sym.isModuleClass)
+        genRegisterReflectiveInstantiationForModuleClass(sym)
+      else
+        genRegisterReflectiveInstantiationForNormalClass(sym)
+    }
+
+    private def genRegisterReflectiveInstantiationForModuleClass(sym: Symbol)(
+        implicit pos: Position): List[js.Tree] = {
+      val fqcnArg = js.StringLiteral(sym.fullName + "$")
+      val runtimeClassArg = js.ClassOf(toReferenceType(sym.info))
+      val loadModuleFunArg = js.Closure(Nil, Nil, genLoadModule(sym), Nil)
+
+      val stat = genApplyMethod(
+          genLoadModule(ReflectModule),
+          Reflect_registerLoadableModuleClass,
+          List(fqcnArg, runtimeClassArg, loadModuleFunArg))
+
+      List(stat)
+    }
+
+    private def genRegisterReflectiveInstantiationForNormalClass(sym: Symbol)(
+        implicit pos: Position): List[js.Tree] = {
+      val ctors =
+        if (sym.isAbstractClass) Nil
+        else sym.info.member(nme.CONSTRUCTOR).alternatives.filter(_.isPublic)
+
+      if (ctors.isEmpty) {
+        Nil
+      } else {
+        val constructorsInfos = for {
+          ctor <- ctors
+        } yield {
+          withNewLocalNameScope {
+            val (parameterTypes, formalParams, actualParams) = (for {
+              param <- ctor.tpe.params
+            } yield {
+              /* Note that we do *not* use `param.tpe` entering posterasure.
+               * Logic would tell us that we should do so, but we intentionally
+               * do not to preserve "bug"-compatibility with the JVM regarding
+               * value classes.
```

Comment:
It applies to all of them. I'll make that clearer.

The "bug" is that if you have:
```scala
class ValueClass(val underlying: Int) extends AnyVal

class Foo(val vc: ValueClass)
```
and you try to instantiate it with reflection, you need to ask for a constructor with a parameter of type `Int` (the underlying type of `ValueClass`), and of provide an actual argument of type `Int(eger)`. This breaks the abstraction of value classes.

This is a recurring theme with value classes, where "reflection" is involved. For example, reflective proxies work with the underlying types of value classes, even though they have a boxed context.